streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -227,6 +255,8 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
       case DeregisterReceiver(streamId, message, error) =>
         deregisterReceiver(streamId, message, error)
         context.reply(true)
+      case ScheduleReceiver(streamId) =>
```

Comment:
Actually, we can merge the change you did in #6294, where `ReceiverSupervisor.onReceiverStart()` (that is, `RegisterReceiver` message) is called before the receiver is started, and the receiver is only started if `RegisterReceiver` message returns `true`. This mechanism can be used by the `ReceiverTracker` to prevent the receiver from starting on unwanted executors - when `RegisterReceiver` is received by the tracker, it will return `!stopping && scheduleReceiver(receiverId).contains(host))`. 

This will solve the race condition as well as check for location without introducing additional messaging... isnt it?